### ISSUE TITLE
Priority or direction in `suggest_trial` for numeric and categorical parameters for biased optimization?

### ISSUE THREAD

**Issue by ayaanhossain** [optuna/optuna#5550]
### Motivation

The hyperparameter optimization problem can sometimes benefit from insights about the nature of a particular hyperparameter. For example, in `Ridge` regression as implemented in `scikit-learn`, the `alpha` parameter controls the regularization on the model coefficients, and ideally should be as large as possible while maximizing cross-validation scores. Currently, there is no way to program or "nudge" in this notion of "maximizing" (or "minimizing") a parameter within an `optuna` `trial` object for a parameter of interest.

### Description

For example, we can do the following
```python
alpha = trial.suggest_float("alpha", 1e-2, 1e3, log=True)
```

but not this
```python
alpha = trial.suggest_float("alpha", 1e-2, 1e3, log=True, direction="maximize")
```
which will allow the user to guide the optimizer in a biased (and useful) way leading to more robust hyperparameters.

This idea of "nudging" can further be extended to categorical hyperparameters by providing a weight to each of them, such that for `SVM` models in `scikit-learn` we can do the following.
```python
kernel = trial.suggest_categorical("kernel", [("linear", 0.5), ("poly", 0.1), ("rbf", 0.4)])
```
which explicitly conveys the user's wish to more frequently sample linear kernels (at least early on) during the hyperparameter optimization loop.

Perhaps, there is already an alternative in the library that I have overlooked (if so, please let me know), but otherwise this looks like a useful addition to the library.

**Comment by nzw0301** [comment]
Thank you for creating this feature request. To be honest, I'm not familiar with this approach. Could you share research papers about this biased optimisation?

**Comment by ayaanhossain** [comment]
Hmm, I don't know if there are any research papers on this. I think what I'm proposing here is more "quality of life" than anything else.

Suppose you have an objective function with multiple variables. You might want to not only optimize the objective function, but also impose constraints on the variables of that function. It's a similar idea, but not exactly the same.

In general, for hyperparameters such as regularization terms like `C` in `SVM` or tree depth in the case of `DecisionTree`, you might want to not only find good values ​​that optimize the performance metric (e.g. `F1` score) of a model trained on a validation dataset, but also minimize those parameters. For example, if both `C=0.01` and `C=0.001` result in an `F1` score of 0.8, I prefer to consider the value of `C=0.001` as the optimal hyperparameter rather than `C=0.01`, since the former has a larger margin and therefore probably results in a more general model.  There is currently no way to ask `optuna` to take this fact into account.

One roundabout way would be to return the hyperparameters themselves as part of the objective function evaluation along with the main objective function evaluation, and make the selected hyperparameter itself one of the values ​​optimized in `optuna` `study` by attaching a `direction` to this extra return value. I think it would be easier to handle this on a "per-parameter basis", though. I don't see how the "hack" I described would take into account the weighting of categorical options, however.

**Comment by nabenabe0928** [comment]
A famous solution is [user-prior](https://arxiv.org/abs/2204.11051).
This algorithm explicitly biases the search to a specific direction.

Another solution would be to combine parameter transformation using Beta distribution.
This solution changes the geometry of the search space and will force some parts of search space to be sampled.

For example, 

```python
alpha = trial.suggest_float("alpha", 1e-2, 1e3, log=True)
```

can be re-written as:

```python
from scipy.stats import beta


a, b = 1, 10  # More weight on the larger side
log_alpha_low = -2
log_alpha_high = 3
alpha_factor = trial.suggest_float("log_alpha_factor", 0, 1)
log10_alpha = log_alpha_low + (log_alpha_high - log_alpha_low) * beta.ppf(alpha_factor, a, b)
alpha = 10 ** log10_alpha
```

This search space increases the probability of yielding higher `alpha`.

However, since the second solution simply distorts the search space, we might experience a situation where we would not get parameter values from a wished range.
Plus, it also adds additional complexity to the search space in the form of the search space transformation parameters.

**Comment by ayaanhossain** [comment]
Thanks for the examples @nabenabe0928! I am happy to close this issue, as your examples solves both of the problems stated regarding directional numeric sampling and weighted categorical sampling.
